### ISSUE TITLE
API does not make much sense

### ISSUE THREAD

**Issue by radio-miskovice** [WICG/serial#111]
Hello, 
I am trying to experiment with the serial API and find it rather disappointing experience.
Problems encountered:
1.  cannot read list of available ports in the system. In my app, I need to show the list of ports to the user and let them pick the correct one to use. Using the built-in dialog (which looks horrible, by the way) is not the way I need to do it in my app.
2. when testing in HTML page from file://... URL, getPorts() always returns empty list. I do not understand the rationale behind this, when local page wants to access local resource and it is just impossible... makes no sense.
3. similar problem with `reader.read()` - the device is sending data back, but those are never read and the respective routine remains in an infinite loop. 
4. Information provided about ports (SerialPort.getInfo()) does not return enough information to keep any kind of persistent information for subsequent use of the same ports. For instance, there is Arduino device (as telegraphy keyer) and a CAT interface, both using the same chip as USB/serial interface. Both have signature 0x0403:0x6001 and there is no way to find any distinguishing feature: neither the system comm port number (COM3 or COM4), nor serial number (in my case A900VLOI and A5XK3RJT, respectively). Coming back to my point 1, I could only read the list of devices *after* they are accessed. Well, if I have to invoke the horrible dialog window and click the potentially needed ports first (one by one!), why not do this at once in the beginning and save time and click-effort? However, without the key identifiers, how can I make any decisions?

I am not willing to take part in the development, but I believe if you look at the issues, you will know what to do.
Thank you for consideration.
Jindra

**Comment by monteslu** [comment]
The hardware APIs prompt for permission before providing system information to help prevent easy user agent fingerprinting.

Use a local webserver for development.  file:// isn't a valid protocol for most secure APIs.  http://localhost works for develop.  All other hosts need to be httpS

The .read() function returns a promise that resolves to an object with a .value property with the data.  You can gate your infinite loop of reads with a check on a port being locked, or whatever makes sense.

I'm sure there's room for improvement on your 4th point, but you'll still need to authorize the port with a prompt before getting info.

**Comment by radio-miskovice** [comment]
The read function returns promise, but the promise never resolves... That was my point. When I connect to the same device, e.g. via putty, I can see the response. Sending works, the keyer is keying CW text when I send it from the web app.

**Comment by radio-miskovice** [comment]
One more remark to the list returned by `getPorts()`: there is no identification whatsoever, just `readable` and `writable` properties (both `null`, which is also basically useless until the port is open). So even when I have the list of ports (previously permitted), I have no information available that I could use to select one or another for my specific peripheral function. They both (or all, if I had more than two) look absolutely identical and none of them has any ID, not even the usbVendorId and usbproductId (also useless - because both are identical, but at least different hardware would look different). This is so absolutely useless...
I give up, I will launch a nodejs process aside, use it as network service, and use the serialport module in nodejs - that's something that was developed by an obviously intelligent and competent person who knows what is serial port used for and what information is necessary to deal with it.

**Comment by reillyeon** [comment]
I would like to help you develop this application however your disrespectful tone and personal attacks in this thread are a violation of the [W3C Code of Ethics and Professional Conduct](https://www.w3.org/Consortium/cepc/) under which this group works. As a fellow amateur radio operator I also find this behavior distasteful and an offense to the hobby.

To your original questions:

1. As Luis mentioned it is a requirement to avoid passive fingerprinting attacks that the user select the serial device they want the site to be able to access.
2. This is a known bug in Chromium related to how permissions are stored internally and tracked as [crbug/1039599](https://crbug.com/1039599) as it applies to more than just this API.
3. It would be helpful to see the code you are using to read from the device.
4. As an author of this specification I am aware of the current limitations of the `getInfo()` method and there are tracking issues open to resolve them while balancing the constraints involved.

If you would like to continue this discussion I am happy to work with you as long as it remains constructive.

**Comment by radio-miskovice** [comment]
In my app I so far do two very simple tasks:
1. send text to the keyer - which works fine
```javascript
const Serial = navigator.serial

let spiderKeyer = undefined
let connected = false

const textbox = document.getElementById("text")
const connectBtn = document.getElementById("connect")

... 

const setupPort = async (port) => {
  spiderKeyer = port
  spiderKeyer.onconnect = (e) => { console.log('port connect: ', e) } // this never happens
  spiderKeyer.ondisconnect = (e) => { console.log('port disconnect: ', e) } // this never happens, too
  const portInfo = await spiderKeyer.getInfo()
  console.log(portInfo) // this works
  spiderKeyer.open({ baudRate: 57600 }).then(x => {
    connected = true
    connectBtn.innerHTML = 'Disconnect'
    connectBtn.onclick = disconnectPort // function omitted, not important for this example
    window.setTimeout(getSpiderId, 2000) // this will later on fire model inquiry - I leave some time for Arduino to initialize
  })
    .catch(e => setError(e))
}

const connectPort = async (e) => {
  if (Serial) {
    if (spiderKeyer) {
      if (!connected) setupPort(spiderKeyer)
    }
    else {
      try {
        const port = await Serial.requestPort()
        setupPort(port)
      }
      catch (e) {
        console.log(e)
      }
    }
    e.preventDefault()
  }
}
connectBtn.onclick = connectPort
textbox.onkeypress = async (e) => {
  if (e.code === 'Enter') {
    e.preventDefault()
    if (!e.ctrlKey && !e.shiftKey && !e.metaKey) {
      let cw = e.target.value
      cw = cw.toUpperCase()
      e.target.value = ''
      console.log(connected ? '' : 'not ', 'sending: ', cw)
      if (spiderKeyer && connected) {
        const w = spiderKeyer.writable.getWriter()
        const t = (new TextEncoder()).encode(cw)
        w.write(t).then(() => { w.releaseLock() }).catch(e => console.log(e))
      }
    }
  }
}

```
The `textbox` element is `<input type="text" id="text" />`, the button is `<button id="connect">Connect Spider Keyer</button>` 

This part works fine, i.e. opening the port and sending text. 

2. What does not work is the attempt to read Spider Keyer version info, which is a sequence of ASCII characters ending with ']' (it is not a terminator, it is the last character of the version string enclosed in brackets).

The following is another part of the same script:
```javascript
// this function should be called asynchonously 
// when the timeout set in setupPort(port) fires (approx. 2 seconds after opening the port)
async function getSpiderId() {
  if (spiderKeyer && connected) {
    let writer = spiderKeyer.writable.getWriter()
    let bf = new ArrayBuffer([17, 17]) // Spider Keyer command "get textual info" [17, anything] http://ok1fig.nagano.cz/SpiderKeyer/SpiderKeyerSpecs.htm 
    await writer.write(bf)
    await writer.releaseLock() // no errors here
    readSpiderId() // switch to read the expected response
  }
}

// reading text response to command [17,x] from Spider Keyer
async function readSpiderId() {
  const reader = spiderKeyer.readable.getReader()
  try {
    console.log('before entering read loop') // this message is printed in the console
    while (true) {
      console.log('now waiting for serial input') // this as well
      let { value, done } = await reader.read()  // this never occurs
    // the rest of the lines in this function never get executed
      if (done) {
        console.log('read loop ended by "done"')
        break;
      }
     // spiderIdText += String.fromCharCode(value) // not important
      if( value === ']') {
        console.log('SPIDER:', spiderIdText)
        break
      }
    }
  }
  catch (e) {
    console.log('error in read loop')
    console.log(e)
  }
  finally {
    console.log( 'finished read loop')
    reader.releaseLock()
    const spiderIdLabel = document.getElementById('status')
    spiderIdLabel.innerText = spiderIdText
    spiderIdText = ''
  }
}
```
As a result of launching `readSpiderId()`, any further attempt to disconnect the port fails. The reader is locked, cannot be released (attempt to `spiderKeyer.readable.getReader().releaseLock()` causes exception 

```
Uncaught (in promise) TypeError: Failed to execute 'getReader' on 'ReadableStream': ReadableStreamReader constructor can only accept readable streams that are not yet locked to a reader
    at HTMLButtonElement.disconnectPort (skpanel.js:15)
```

I could perhaps try to make the `reader` variable global and then somehow force it to `releaseLock()` but I think it will just cause another exception, complaining something like the stream is not finished or something similar (it was in another version of the code which I did not keep).

Since the delay before the reading is started is 2 seconds, when I click the Disconnect button early enough after Connect, before `getSpiderId()` is called, the port is disconnected and then I can connect again. But once the reader is activated, I cannot disconnect by any means except reload the page. Sending still works, though. 

I appreciate you are trying to help, but my experience with this API has been mostly frustrating and very time consuming, and unfortunately it is not the solution for the task I want to accomplish. I hope you understand.

73 Jindra

**Comment by radio-miskovice** [comment]
OK, understood, I apologize.

My app design goes far beyond this topic - in the meantime I found a bug that caused the reader to hang and the remaining issues will eventually better fit as question(s) to stackoverflow.com, as they are rather issues in my understanding of the underlying objects rather than the API itself. 

To stay on-topic, regarding the port identification, may I suggest that some basic identification be already included even in the result of `navigator.serial.getPorts()`. Otherwise each port in the list is identified only by its internal address or ID (not accessible by the javascript program, unless I am mistaken), so at least the system comm port name would be helpful, and for USB devices their serial number (if available) would be even more helpful. If these where not included, the program would have to call an extra *port*.`getInfo()` for each port before it would be possible to check, for instance, if that is the expected device for the given task. This identification is inevitable, so I think why not to include the information right away.

**Comment by reillyeon** [comment]
Let's keep discussion of adding properties to the object returned by `getInfo()` on #19 as well as the corresponding Chromium [issue 1074383](https://crbug.com/1074383) which collects more motivation from developers for exposing more device properties.

I see you posted and then deleted your example code. I took a look and it seems correct based on the documentation for your device. This may indicate that you have found a bug in the implementation. If you can post the code to a new issue on [crbug.com](https://crbug.com/new) we can continue this discussion as an implementation rather than specification issue.

**Comment by radio-miskovice** [comment]
Yes, I spent about an hour composing a detailed description and because I had to look for pieces of code to be pasted in, I looked carefully into one of the functions where I found a stupid bug. The function sent empty buffer (0 bytes) instead of the 2-byte command it was supposed to send, hence no response. So it made no sense to post anything... ;-)
There are some things to consider but those are unrelated to the topic. The debugged first version of the app can be seen here: https://youtu.be/trP3q7OR9GI it's pretty rough, there are still no controls (wpm, pitch), it's a toy project to learn the API.
